Replace `AccountSettingsService.context` with `CoreDataStack`

crazytonyli opened the pull request:
Use `CoreDataStack` instead of `NSManagedObjectContext` in `AccountSettingsService`. See the "Issue" and "Proposal" sections of https://github.com/wordpress-mobile/WordPress-iOS/pull/19893 for rational behind this change.

## Test instructions

Tap the user avatar icon, go to "Account Settings" page, verify all settings can be changed.

## Regression Notes
1. Potential unintended areas of impact
None.

2. What I did to test those areas of impact (or what existing automated tests I relied on)
None.

3. What automated tests I added (or what prevented me from doing so)
A couple of unit tests to verify changes are saved or reverted correctly.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

dependabot[bot] committed:
Bump cocoapods-downloader from 1.5.1 to 1.6.3

Bumps [cocoapods-downloader](https://github.com/CocoaPods/cocoapods-downloader) from 1.5.1 to 1.6.3.
- [Release notes](https://github.com/CocoaPods/cocoapods-downloader/releases)
- [Changelog](https://github.com/CocoaPods/cocoapods-downloader/blob/master/CHANGELOG.md)
- [Commits](https://github.com/CocoaPods/cocoapods-downloader/compare/1.5.1...1.6.3)

---
updated-dependencies:
- dependency-name: cocoapods-downloader
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>

crazytonyli committed:
Replace `AccountSettingsService.context` with `CoreDataStack`

crazytonyli committed:
Add unit tests for `AccountSettingsService`

crazytonyli committed:
Update release notes

mokagio commented on WordPress/Classes/Utility/CoreDataHelper.swift:
```diff
@@ -201,6 +201,19 @@ extension CoreDataStack {
+    /// which is available from iOS 15.0 (hence the deprecation attribute below).
+    @available(iOS, deprecated: 15.0, message: "Use `mainContext.performAndWait` instead.")
+    func performQuery<T>(_ block: @escaping (NSManagedObjectContext) -> T) -> T {
+        var value: T! = nil
```
At first sight, I didn't make sense to me. I though, "_How can assigning `nil` to a IUO work?!_" Then, I remembered reading about some under the hood changes to IUO some time ago and found the [this article on `swift.org`](https://www.swift.org/blog/iuo/). 

> Implicitly unwrapped optionals have been reimplemented such that they are no longer a distinct type from `Optional<T>`. As a result, type checking is more consistent and there are fewer special cases in the compiler. Removing these special cases should lead to fewer bugs in handling of these declarations.

No that makes more sense 😄 

Trying the code in the REPL confirms it:

<img width="1165" alt="image" src="https://user-images.githubusercontent.com/1218433/213033113-101c9487-de5e-4898-85aa-b0796f35c62d.png">

jkmassel replied:
Given the fact that this method is defined on the stack, and not on `ManagedObjectContext`, I wonder if we should deprecate it after all? 

It can still be useful for isolating the underlying `mainContext` if that's what we're after? Otherwise, should we define it as an extension on `ManagedObjectContext` to make the refactor more straightforward later?

crazytonyli replied:
> It can still be useful for isolating the underlying mainContext if that's what we're after?

Agreed. I think it's nice to have `performQuery` for reading and `performAndSave` for writing. I've removed the deprecation attribute in https://github.com/wordpress-mobile/WordPress-iOS/pull/19910/commits/139cc5789b153b9110a0f3a3f1c8b07b75c60a97

mokagio commented on WordPress/Classes/Utility/CoreDataHelper.swift:
```diff
@@ -201,6 +201,19 @@ extension CoreDataStack {
+    @available(iOS, deprecated: 15.0, message: "Use `mainContext.performAndWait` instead.")
+    func performQuery<T>(_ block: @escaping (NSManagedObjectContext) -> T) -> T {
+        var value: T! = nil
+        self.mainContext.performAndWait {
```
Nitpick: `self` should not be necessary, right?

```suggestion
        mainContext.performAndWait {
```

jkmassel replied:
If we want to enforce "no unneeded `self` references", we should add that rule to SwiftLint!

mokagio replied:
Great point!

mokagio commented on WordPress/WordPressTest/AccountSettingsServiceTests.swift:
```diff
@@ -0,0 +1,74 @@
+
+@testable import WordPress
+
+class AccountSettingsServiceTests: CoreDataTestCase {
```
Yes for more tests!

jkmassel reviewed (dismissed):
Had a question about `performQuery` that might be nice to sort out before merging, but don't consider it strictly required

crazytonyli committed:
Remove the depreacation attribute on `performQuery`

crazytonyli committed:
Merge branch 'trunk' into remove-context-from-account-settings-service

crazytonyli committed:
Merge branch 'trunk' into remove-context-from-account-settings-service

alpavanoglu committed:
Add empty block to prompts editor

hassaanelgarem committed:
Add: functions for light and dark variants of UIColor

hassaanelgarem committed:
Update: Change color of the close button in overlays

alpavanoglu committed:
Add Feature flag condition

dependabot[bot] committed:
Bump commonmarker from 0.23.6 to 0.23.7

Bumps [commonmarker](https://github.com/gjtorikian/commonmarker) from 0.23.6 to 0.23.7.
- [Release notes](https://github.com/gjtorikian/commonmarker/releases)
- [Changelog](https://github.com/gjtorikian/commonmarker/blob/main/CHANGELOG.md)
- [Commits](https://github.com/gjtorikian/commonmarker/compare/v0.23.6...v0.23.7)

---
updated-dependencies:
- dependency-name: commonmarker
  dependency-type: direct:production
...

Signed-off-by: dependabot[bot] <support@github.com>

crazytonyli committed:
Merge branch 'trunk' into remove-context-from-account-settings-service

crazytonyli committed:
Merge branch 'trunk' into remove-context-from-account-settings-service

crazytonyli committed:
Update WordPressKit

alpavanoglu committed:
Merge pull request #19967 from wordpress-mobile/feature/prompts-empty-paragraph-block

[Blogging prompts] Add empty block below blogging prompts block

jkmassel committed:
Merge pull request #19313 from wordpress-mobile/dependabot/bundler/cocoapods-downloader-1.6.3

Bump cocoapods-downloader from 1.5.1 to 1.6.3

jkmassel committed:
Merge pull request #19988 from wordpress-mobile/dependabot/bundler/commonmarker-0.23.7

Bump commonmarker from 0.23.6 to 0.23.7

crazytonyli committed:
Merge pull request #19908 from wordpress-mobile/tonyli-wordpresskit-model-iuo

Update WordPressKit

hassaanelgarem committed:
Merge pull request #19970 from wordpress-mobile/task/19599-overlay-button-background

Jetpack Focus: Update the overlay's close button background in dark mode

crazytonyli committed:
Update WordPressShared to the latest version

It was accidentally reverted in https://github.com/wordpress-mobile/WordPress-iOS/pull/19908

crazytonyli committed:
Merge pull request #19998 from wordpress-mobile/tonyli-wordpresskit-model-iuo

Update WordPressShared to the latest version

crazytonyli committed:
Merge pull request #19910 from wordpress-mobile/remove-context-from-account-settings-service

Replace `AccountSettingsService.context` with `CoreDataStack`

mokagio commented:
Hey @crazytonyli, from the sequence of events, I'm guessing this PR was closed intentionally. Is that right?

crazytonyli commented:
I believe the changes were merged, but somehow GitHub mis-labled this PR as closed.

The "Reopen and comment" button says "These commits are already" merged
![Untitled](https://user-images.githubusercontent.com/1101828/215358002-c32de67c-2ef4-493f-9282-882771046393.png)

And, https://github.com/wordpress-mobile/WordPress-iOS/commit/c58826f14d30b6e2f39b3365f0dae0195fe81aa9 is indeed in the trunk branch
<img width="545" alt="Screenshot 2023-01-30 at 10 59 01 AM" src="https://user-images.githubusercontent.com/1101828/215357913-f7d15c5a-1443-4634-965c-3a47598aa21a.png">

mokagio commented:
Ah, okay. No worries 😄
